presto-main/src/main/java/com/facebook/presto/operator/ThriftRpcShuffleClient.java

```diff
@@ -0,0 +1,94 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.operator;
+
+import com.facebook.airlift.log.Logger;
+import com.facebook.drift.client.DriftClient;
+import com.facebook.presto.execution.TaskId;
+import com.facebook.presto.execution.buffer.BufferResult;
+import com.facebook.presto.execution.buffer.OutputBuffers.OutputBufferId;
+import com.facebook.presto.operator.PageBufferClient.PagesResponse;
+import com.facebook.presto.server.thrift.ThriftTaskClient;
+import com.google.common.util.concurrent.Futures;
+import com.google.common.util.concurrent.ListenableFuture;
+import io.airlift.units.DataSize;
+
+import javax.annotation.concurrent.ThreadSafe;
+
+import java.net.URI;
+import java.util.Optional;
+import java.util.concurrent.Executor;
+
+import static com.facebook.presto.operator.PageBufferClient.PagesResponse.createPagesResponse;
+import static com.google.common.util.concurrent.MoreExecutors.directExecutor;
+import static java.util.Objects.requireNonNull;
+
+@ThreadSafe
+public final class ThriftRpcShuffleClient
+        implements RpcShuffleClient
+{
+    private static final Logger log = Logger.get(ThriftRpcShuffleClient.class);
+
+    private final ThriftTaskClient thriftClient;
+    private final Executor executor;
+    private final TaskId taskId;
+    private final OutputBufferId outputBufferId;
+
+    public ThriftRpcShuffleClient(DriftClient<ThriftTaskClient> driftClient, URI location, Executor executor)
+    {
+        requireNonNull(location, "location is null");
+
+        this.thriftClient = requireNonNull(driftClient, "thriftClient is null").get(Optional.of(location.getAuthority()));
+        this.executor = requireNonNull(executor, "executor is null");
+
+        // TODO: refactor the entire LocationFactory interfaces to completely replace URI with more efficient/expressive data structures
+        String[] paths = location.getPath().split("/");
+        this.taskId = TaskId.valueOf(paths[3]);
```

Comment:
Before we address the TODO, let's leave a comment here about the structure of the paths so that it is easier to perceive how this magic 3 and 5 works.